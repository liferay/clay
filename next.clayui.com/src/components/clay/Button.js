/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import ClayButton from '@clayui/button';
import Editor from './Editor';
import React from 'react';

const spritemap = '/images/icons/icons.svg';

const ButtonDisplayTypesCode = `const Component = () => {
	return (
		<>
			<ClayButton displayType="primary">
				Button Primary
			</ClayButton>
			<ClayButton displayType="secondary">
				Button Secondary
			</ClayButton>
			<ClayButton displayType="link">
				Button Link
			</ClayButton>
			<ClayButton displayType="unstyled">
				Button Unstyled
			</ClayButton>
		</>
	);
}

render(<Component />);`;

const ButtonDisplayTypes = () => {
	const scope = {ClayButton, spritemap};
	const code = ButtonDisplayTypesCode;

	return <Editor code={code} scope={scope} />;
};

const ButtonGroupCode = `const Component = () => {
	return (
        <ClayButton.Group>
			<ClayButton>{'This'}</ClayButton>
			<ClayButton displayType="secondary">{'is'}</ClayButton>
			<ClayButton>{'a'}</ClayButton>
			<ClayButton displayType="secondary">{'button'}</ClayButton>
			<ClayButton>{'group.'}</ClayButton>
		</ClayButton.Group>
	);
}

render(<Component />);`;

const ButtonGroup = () => {
	const scope = {ClayButton, spritemap};
	const code = ButtonGroupCode;

	return <Editor code={code} scope={scope} />;
};

export {ButtonDisplayTypes, ButtonGroup};
