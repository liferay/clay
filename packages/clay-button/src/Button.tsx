/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import classNames from 'classnames';
import React from 'react';

import ButtonGroup from './Group';

export type DisplayType = 'primary' | 'secondary' | 'link' | 'unstyled';

interface IProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
	/**
	 * Flag to indicate if link should be borderless.
	 */
	borderless?: boolean;

	/**
	 * Renders the button as a block element.
	 */
	block?: boolean;

	/**
	 * Determines how to button is displayed.
	 */
	displayType?: DisplayType;

	/**
	 * Flag to indicate if button should be monospaced.
	 */
	monospaced?: boolean;

	/**
	 * Flag to indicate if link need have an outline.
	 */
	outline?: boolean;

	/**
	 * Indicates button should be a small variant.
	 */
	small?: boolean;
}

type Button = React.ForwardRefExoticComponent<
	IProps & React.RefAttributes<HTMLButtonElement>
> & {Group: typeof ButtonGroup};

const ClayButton = React.forwardRef(
	(
		{
			block,
			borderless,
			children,
			className,
			displayType = 'primary',
			monospaced,
			outline,
			small,
			type = 'button',
			...otherProps
		}: IProps,
		ref
	) => (
		<button
			className={classNames(className, 'btn', {
				'btn-block': block,
				'btn-monospaced': monospaced,
				'btn-outline-borderless': borderless,
				'btn-sm': small,
				[`btn-${displayType}`]: displayType && !outline && !borderless,
				[`btn-outline-${displayType}`]:
					displayType && (outline || borderless),
			})}
			ref={ref}
			type={type}
			{...otherProps}
		>
			{children}
		</button>
	)
) as Button;

ClayButton.Group = ButtonGroup;

export default ClayButton;
