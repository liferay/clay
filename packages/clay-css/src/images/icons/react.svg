<!--
* SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
* SPDX-FileCopyrightText: © 2020 Contributors to the project Clay <https://github.com/liferay/clay/graphs/contributors>
*
* SPDX-License-Identifier: BSD-3-Clause
-->
<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg"><path d="M16 8.006c0-1.041-1.327-2.028-3.362-2.64.47-2.038.26-3.66-.66-4.179a1.438 1.438 0 0 0-.726-.18v-.003c-.897 0-2.071.628-3.259 1.718-1.187-1.083-2.36-1.705-3.258-1.705v.003c-.274 0-.521.058-.737.18-.916.519-1.122 2.137-.649 4.169C1.321 5.984 0 6.968 0 8.006c0 1.041 1.327 2.028 3.362 2.64-.47 2.038-.26 3.66.66 4.179.211.122.459.18.733.18.897 0 2.07-.629 3.258-1.718 1.187 1.083 2.361 1.705 3.258 1.705.274 0 .522-.058.737-.18.917-.519 1.122-2.137.65-4.169C14.678 10.031 16 9.044 16 8.006Zm-4.752-6.287c.15.003.271.032.372.086.444.25.636 1.202.486 2.426a9.73 9.73 0 0 1-.166.942 16.066 16.066 0 0 0-2.071-.349 15.698 15.698 0 0 0-1.357-1.602c1.06-.971 2.058-1.5 2.736-1.503ZM5.11 9.647c.154.256.31.506.47.75a14.93 14.93 0 0 1-1.337-.254c.12-.413.27-.84.44-1.265.134.256.274.512.427.769Zm-.877-3.778c.427-.1.877-.186 1.344-.253a17.649 17.649 0 0 0-.894 1.528 13.57 13.57 0 0 1-.45-1.275Zm.832 2.147a17.028 17.028 0 0 1 1.461-2.503 17.267 17.267 0 0 1 2.942-.003 16.59 16.59 0 0 1 1.468 2.493 17.032 17.032 0 0 1-1.461 2.503 17.251 17.251 0 0 1-2.943.003 16.59 16.59 0 0 1-1.468-2.493Zm5.824 1.621c.15-.256.29-.516.425-.772.176.43.326.859.45 1.275-.428.103-.878.19-1.344.256.16-.246.32-.5.47-.759Zm-.003-3.272a17.37 17.37 0 0 0-.47-.75c.464.068.91.151 1.338.254-.12.413-.27.84-.44 1.265a15.136 15.136 0 0 0-.428-.769Zm-1.99-1.611A20.543 20.543 0 0 0 8 4.73c-.307 0-.61.007-.907.023.294-.375.597-.718.897-1.026.304.308.607.65.907 1.026ZM4.36 1.818a.736.736 0 0 1 .375-.083v-.003c.676 0 1.677.529 2.74 1.493a15.197 15.197 0 0 0-1.347 1.599c-.737.077-1.435.196-2.074.353a9.516 9.516 0 0 1-.17-.93c-.153-1.224.036-2.176.476-2.429Zm-.82 8.136c-.323-.1-.63-.208-.914-.327C1.471 9.144.724 8.51.724 8.006c0-.503.747-1.14 1.901-1.621.28-.119.587-.225.904-.324.186.628.43 1.282.734 1.951-.3.667-.542 1.317-.724 1.942Zm1.21 4.339a.79.79 0 0 1-.37-.086c-.443-.25-.635-1.202-.485-2.426.036-.301.095-.619.166-.942.64.153 1.337.272 2.071.349.44.593.897 1.131 1.357 1.602-1.063.971-2.061 1.503-2.74 1.503Zm3.258-2.01A13.3 13.3 0 0 1 7.1 11.26c.293.012.593.022.897.022.306 0 .61-.007.906-.023-.293.375-.596.718-.896 1.026Zm3.633 1.91a.736.736 0 0 1-.375.084c-.676 0-1.677-.529-2.74-1.493.457-.471.913-1.006 1.347-1.6a15.478 15.478 0 0 0 2.074-.352c.075.324.134.635.17.933.153 1.224-.036 2.175-.476 2.429Zm1.731-4.566c-.28.119-.587.225-.903.324A15.22 15.22 0 0 0 11.734 8c.3-.667.542-1.317.724-1.942.323.1.63.208.917.327 1.154.483 1.901 1.118 1.901 1.621-.003.503-.75 1.14-1.905 1.621Z"/><path d="M7.997 9.47c.823 0 1.49-.655 1.49-1.464 0-.809-.667-1.464-1.49-1.464s-1.49.655-1.49 1.464c0 .809.667 1.464 1.49 1.464Z"/></svg>
