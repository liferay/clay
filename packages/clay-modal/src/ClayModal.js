import 'clay-button';
import 'clay-icon';
import {Config} from 'metal-state';
import {EventHandler} from 'metal-events';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import dom from 'metal-dom';
import Soy from 'metal-soy';

import templates from './ClayModal.soy.js';

const KEY_CODE_ESC = 27;

/**
 * Metal ClayModal component.
 */
class ClayModal extends Component {
	/**
	 * @inheritDoc
	 */
	created() {
		this._overlayElement = this._valueOverlayElementFn();
		this._eventHandler = new EventHandler();
	}

	/**
	 * @inheritDoc
	 */
	attached() {
		this.addListener('click', this._handleDocumentClick.bind(this), true);
		this.addListener('hide', this._defaultHideModal, true);
		this.addListener('show', this._defaultShowModal, true);
		this.addListener(
			'touchend',
			this._handleDocumentClick.bind(this),
			true
		);
		this.addListener('transitionend', this._handleTransitionEnd, true);
	}

	/**
	 * @inheritDoc
	 */
	detached() {
		super.detached();
		this._eventHandler.removeAllListeners();
	}

	/**
	 * @inheritDoc
	 */
	// eslint-disable-next-line
	sync_isTransitioning() {
		if (this._isTransitioning && !this.visible) {
			this._isTransitioning = false;
			this.visible = true;
		} else if (this._isTransitioning && this.visible) {
			this.visible = false;
		}
	}

	/**
	 * @inheritDoc
	 */
	syncVisible() {
		if (this.visible) {
			dom.enterDocument(this._overlayElement);
			this._overlayElement.offsetHeight;
			dom.addClasses(this._overlayElement, 'show');
		} else {
			dom.exitDocument(this._overlayElement);
		}
	}

	/**
	 * Close modal and remove transitions.
	 * @private
	 */
	_defaultHideModal() {
		dom.removeClasses(this._overlayElement, 'show');
		dom.removeClasses(document.body, 'modal-open');

		this._isTransitioning = true;

		this._eventHandler.removeAllListeners();
	}

	/**
	 * Open modal and add transition.
	 * @private
	 */
	_defaultShowModal() {
		dom.addClasses(document.body, 'modal-open');

		this._isTransitioning = true;

		this._eventHandler.add(
			dom.on(document, 'keyup', this._handleKeyup.bind(this))
		);
	}

	/**
	 * Handle the click buttons and emit event.
	 * @param {!Event} event
	 * @private
	 */
	_handleClickFooterButton(event) {
		this.emit('clickButton', event);
	}

	/**
	 * Handle the footer close button click and emits events.
	 * @param {!Event} event
	 * @private
	 */
	_handleClickCloseButtonFooter(event) {
		this.emit('clickButton', event);
		this.emit('hide');
	}

	/**
	 * Handle the click and emit close modal.
	 * @private
	 */
	_handleCloseModal() {
		this.emit('hide');
	}

	/**
	 * Check if contain click target event in element and emit close modal.
	 * @param {!Event} event
	 * @private
	 */
	_handleDocumentClick(event) {
		if (dom.contains(event.target, this.element)) {
			this.emit('hide');
		}
	}

	/**
	 * Handle click key code esc and emit close modal.
	 * @param {!Event} event
	 * @private
	 */
	_handleKeyup(event) {
		if (event.keyCode === KEY_CODE_ESC) {
			this.emit('hide');
		}
	}

	/**
	 * Handle css transition ends.
	 * @param {!Event} event
	 * @private
	 */
	_handleTransitionEnd(event) {
		if (
			event.target === this.element &&
			this._isTransitioning &&
			!this.visible
		) {
			this._isTransitioning = false;
		}
	}

	/**
	 * Create fragment of the overlay modal.
	 * @return {Element} The resulting document fragment.
	 * @private
	 */
	_valueOverlayElementFn() {
		return dom.buildFragment('<div class="modal-backdrop fade"></div>')
			.firstChild;
	}

	/**
	 * Emit open modal.
	 * @public
	 */
	show() {
		this.emit('show');
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayModal.STATE = {
	/**
	 * Body of the modal.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|html|undefined}
	 * @default undefined
	 */
	body: Config.any(),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Buttons of the footer.
	 * @instance
	 * @memberof ClayModal
	 * @type {?array|undefined}
	 * @default undefined
	 */
	footerButtons: Config.arrayOf(
		Config.shapeOf({
			alignment: Config.oneOf(['left', 'right']).value('right'),
			label: Config.string().required(),
			style: Config.oneOf(['link', 'primary', 'secondary', 'unstyled']),
			type: Config.oneOf(['button', 'close', 'reset', 'submit']).value(
				'button'
			),
		})
	),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Adds transitions when show is called.
	 * @instance
	 * @memberof ClayModal
	 * @type {?bool}
	 * @default false
	 * @private
	 */
	_isTransitioning: Config.bool()
		.value(false)
		.internal(),

	/**
	 * The size of element modal.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|undefined}
	 * @default undefined
	 */
	size: Config.oneOf(['full-screen', 'lg', 'sm']),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|undefined}
	 * @default undefined
	 */
	spritemap: Config.string(),

	/**
	 * Status messages.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|undefined}
	 * @default undefined
	 */
	status: Config.oneOf(['danger', 'info', 'success', 'warning']),

	/**
	 * Title of the modal.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|undefined}
	 * @default undefined
	 */
	title: Config.string(),

	/**
	 * Url to place an iframe in the body of the modal.
	 * @instance
	 * @memberof ClayModal
	 * @type {?string|undefined}
	 * @default undefined
	 */
	url: Config.string(),

	/**
	 * Modal visible when show is called.
	 * @instance
	 * @memberof ClayModal
	 * @type {?bool}
	 * @default false
	 * @private
	 */
	visible: Config.bool().value(false),
};

defineWebComponent('clay-modal', ClayModal);

Soy.register(ClayModal, templates);

export {ClayModal};
export default ClayModal;
