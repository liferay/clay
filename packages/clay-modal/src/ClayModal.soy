{namespace ClayModal}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param? _handleClickCloseButtonFooter: any}
	{@param? _handleClickFooterButton: any}
	{@param? _handleCloseModal: any}
	{@param? _isTransitioning: bool}
	{@param? body: string|html}
	{@param? elementClasses: string}
	{@param? footerButtons: list<?>}
	{@param? id: string}
	{@param? size: string}
	{@param? spritemap: string}
	{@param? status: string}
	{@param? title: string}
	{@param? url: string}
	{@param? visible: bool}

	{let $attributes kind="attributes"}
		class="fade modal
			{if $_isTransitioning or $visible}
				{sp}d-block
			{/if}

			{if $visible}
				{sp}show
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}

		role="dialog"
		tabindex="-1"
	{/let}

	<div {$attributes}>
		{let $classesSize kind="text"}
			modal-dialog
			{if $size}
				{sp}modal-{$size}
			{/if}

			{if $status}
				{sp}modal-{$status}
			{/if}
		{/let}

		<div class="{$classesSize}">
			<div class="modal-content">
				{call .header}
					{param _handleCloseModal: $_handleCloseModal /}
					{param spritemap: $spritemap /}
					{param status: $status /}
					{param title: $title /}
				{/call}

				{call .body}
						{param content: $body /}
						{param url: $url /}
				{/call}

				{call .footer}
					{param _handleClickCloseButtonFooter: $_handleClickCloseButtonFooter /}
					{param _handleClickFooterButton: $_handleClickFooterButton /}
					{param buttons: $footerButtons /}
				{/call}
			</div>
		</div>
	</div>
{/template}

/**
 * Render a header of the modal.
 */
{template .header}
	{@param? _handleCloseModal: any}
	{@param? spritemap: string}
	{@param? status: string}
	{@param? title: string}

	{if $title}
		<div class="modal-header">
			{if $spritemap and $status}
				{let $icon kind="text"}
					{switch $status}
						{case 'danger'}
							exclamation-full
						{case 'success'}
							check-circle
						{case 'warning'}
							question-circle-full
						{default}
							info-circle
					{/switch}
				{/let}

				<span class="modal-title-indicator">
					{call ClayIcon.render}
						{param spritemap: $spritemap /}
						{param symbol: $icon /}
					{/call}
				</span>
			{/if}

			<div class="modal-title">{$title}</div>

			{if $spritemap}
				{call ClayButton.render}
					{param elementClasses: 'close' /}
					{param events: ['click': $_handleCloseModal] /}
					{param icon: 'times' /}
					{param spritemap: $spritemap /}
					{param style: 'unstyled' /}
				{/call}
			{/if}
		</div>
	{/if}
{/template}

/**
 * Render a body of the modal.
 */
{template .body}
	{@param? content: string|html}
	{@param? url: string}

	{if $url}
		<div class="modal-body modal-body-iframe">
			{call .iframe}
				{param url: $url /}
			{/call}
		</div>
	{elseif $content}
		<div class="modal-body">
			{$content}
		</div>
	{/if}
{/template}

/**
 * Render a footer of the modal.
 */
{template .footer}
	{@param? _handleClickCloseButtonFooter: any}
	{@param? _handleClickFooterButton: any}
	{@param? buttons: list<?>}

	{if $buttons}
		<div class="modal-footer">
			<div class="modal-item-first">
				{foreach $button in $buttons}
					{if $button.alignment == 'left'}
						{call ClayButton.render}
							{param events: ['click': $_handleClickFooterButton] /}
							{param label: $button.label /}
							{param style: $button.style ?: 'secondary' /}
							{param type: not $button.type or $button.type == 'close' ? 'button' : $button.type /}
						{/call}
						{sp}
					{/if}
				{/foreach}
			</div>
			<div class="modal-item-last">
				{foreach $button in $buttons}
					{if not $button.alignment or $button.alignment == 'right'}
						{if $button.type == 'close'}
							{call ClayButton.render}
								{param elementClasses: 'close-modal' /}
								{param events: ['click': $_handleClickCloseButtonFooter] /}
								{param label: $button.label /}
								{param style: 'secondary' /}
								{param type: 'button' /}
							{/call}
						{else}
							{call ClayButton.render}
								{param events: ['click': $_handleClickFooterButton] /}
								{param label: $button.label /}
								{param style: $button.style ?: 'primary' /}
								{param type: $button.type /}
							{/call}
						{/if}
						{sp}
					{/if}
				{/foreach}
			</div>
		</div>
	{/if}
{/template}

/**
 * Renders an iframe.
 */
{template .iframe}
	{@param url: string}

	<iframe src="{$url}" title="{$url}"></iframe>
{/template}
