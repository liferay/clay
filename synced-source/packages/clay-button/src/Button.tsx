/**
 * SPDX-FileCopyrightText: Â© 2019 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

import LoadingIndicator from '@clayui/loading-indicator';
import classNames from 'classnames';
import React from 'react';
import warning from 'warning';

import Group from './Group';

type DisplayType =
	| null
	| 'primary'
	| 'secondary'
	| 'link'
	| 'success'
	| 'warning'
	| 'danger'
	| 'info'
	| 'beta'
	| 'beta-dark'
	| 'unstyled';

export interface IProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
	/**
	 * Flag to indicate if button is used within an alert component.
	 */
	alert?: boolean;

	/**
	 * Flag to indicate if the button should be borderless.
	 */
	borderless?: boolean;

	/**
	 * Renders the button as a block element.
	 */
	block?: boolean;

	/**
	 * Flag to indicate if the button should use the clay-dark variant.
	 */
	dark?: boolean;

	/**
	 * Determines the button variant to use.
	 * The values `beta` and `beta-dark` are deprecated since v3.100.0 - use
	 * `translucent` and `dark` instead.
	 * The values `null` and `unstyled` are for internal use only.
	 */
	displayType?: DisplayType;

	/**
	 * Flag to display a loading indicator.
	 */
	loading?: boolean;

	/**
	 * Flag to indicate if button should be monospaced.
	 */
	monospaced?: boolean;

	/**
	 * Flag to indicate if the button should use the outline variant.
	 */
	outline?: boolean;

	/**
	 * Flag to indicate if the button should be shaped like a pill.
	 */
	rounded?: boolean;

	/**
	 * Determines the size of a button.
	 */
	size?: 'xs' | 'regular' | 'sm';

	/**
	 * Indicates button should be a small variant.
	 * @deprecated since v3.72.0 - use `size` instead.
	 */
	small?: boolean;

	/**
	 * Flag to indicate if the button should use the translucent variant.
	 */
	translucent?: boolean;
}

export interface IForwardRef<T, P = {}>
	extends React.ForwardRefExoticComponent<P & React.RefAttributes<T>> {
	Group: typeof Group;
}

function forwardRef<T, P = {}>(component: React.RefForwardingComponent<T, P>) {
	return React.forwardRef<T, P>(component) as IForwardRef<T, P>;
}

const Button = forwardRef(
	(
		{
			alert,
			block,
			borderless,
			children,
			className,
			dark,
			displayType = 'primary',
			loading,
			monospaced,
			outline,
			rounded,
			size = 'regular',
			small,
			translucent,
			type = 'button',
			...otherProps
		}: IProps,
		ref: React.Ref<HTMLButtonElement>
	) => {
		const childArray = React.Children.toArray(children);

		warning(
			!(
				childArray.length === 1 &&
				// @ts-ignore
				childArray[0].type?.displayName === 'ClayIcon' &&
				typeof otherProps['aria-label'] !== 'string' &&
				typeof otherProps['aria-labelledby'] !== 'string'
			),
			'Button Accessibility: Component has only the Icon declared. Define an `aria-label` or `aria-labelledby` attribute that labels the interactive button that screen readers can read. The `title` attribute is optional but consult your design team.'
		);

		if (displayType === 'beta') {
			displayType = 'info';
			translucent = true;
		} else if (displayType === 'beta-dark') {
			dark = true;
			displayType = 'info';
			translucent = true;
		}

		return (
			<button
				className={classNames(className, 'btn', {
					'alert-btn': alert,
					'btn-block': block,
					'btn-monospaced': monospaced,
					'btn-outline-borderless': borderless,
					'btn-sm': small && (!size || size === 'regular'),
					'btn-translucent': translucent,
					'clay-dark': dark,
					[`btn-${displayType}`]:
						displayType && !outline && !borderless,
					[`btn-outline-${displayType}`]:
						displayType && (outline || borderless),
					'rounded-pill': rounded,
					[`btn-${size}`]: size && size !== 'regular',
				})}
				ref={ref}
				type={type}
				{...otherProps}
			>
				{loading && (
					<span className="inline-item inline-item-before">
						<LoadingIndicator />
					</span>
				)}
				{children}
			</button>
		);
	}
);

Button.Group = Group;
Button.displayName = 'ClayButton';

export default Button;
