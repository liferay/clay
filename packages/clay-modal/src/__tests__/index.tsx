/*eslint no-console: 0 */
/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import Button from '@clayui/button';
import ClayModal, {useModal} from '..';
import React from 'react';
import ReactDOM from 'react-dom';
import {act, cleanup, render} from '@testing-library/react';

const spritemap = 'icons.svg';

describe('ClayModal', () => {
	beforeAll(() => {
		// @ts-ignore
		ReactDOM.createPortal = jest.fn(element => {
			return element;
		});
		jest.useFakeTimers();
	});

	afterAll(() => {
		jest.useRealTimers();
	});

	afterEach(() => {
		jest.clearAllTimers();
		jest.restoreAllMocks();
		cleanup();
	});

	it('renders', () => {
		const ModalWithState = () => {
			const {observer, onClose} = useModal({onClose: () => {}});

			return (
				<ClayModal observer={observer} spritemap={spritemap}>
					<ClayModal.Header>{'Foo'}</ClayModal.Header>
					<ClayModal.Body>
						<h1>{'Hello world!'}</h1>
					</ClayModal.Body>
					<ClayModal.Footer
						last={<Button onClick={onClose}>{'Primary'}</Button>}
					/>
				</ClayModal>
			);
		};

		render(<ModalWithState />);

		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toMatchSnapshot();
	});

	it('renders with Header', () => {
		const ModalWithState = () => {
			const {observer} = useModal({onClose: () => {}});
			return (
				<ClayModal observer={observer} spritemap={spritemap}>
					<ClayModal.Header>{'Foo'}</ClayModal.Header>
				</ClayModal>
			);
		};

		render(<ModalWithState />);

		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toMatchSnapshot();
	});

	it('renders with size', () => {
		const ModalWithState = () => {
			const {observer} = useModal({onClose: () => {}});
			return (
				<ClayModal
					observer={observer}
					size="lg"
					spritemap={spritemap}
				/>
			);
		};

		render(<ModalWithState />);

		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toMatchSnapshot();
	});

	it('renders with status', () => {
		const ModalWithState = () => {
			const {observer} = useModal({onClose: () => {}});
			return (
				<ClayModal
					observer={observer}
					spritemap={spritemap}
					status="success"
				/>
			);
		};

		render(<ModalWithState />);

		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toMatchSnapshot();
	});

	it('renders a body component with url', () => {
		const ModalWithState = () => {
			const {observer} = useModal({onClose: () => {}});
			return (
				<ClayModal observer={observer} spritemap={spritemap}>
					<ClayModal.Body url="http://localhost" />
				</ClayModal>
			);
		};

		render(<ModalWithState />);

		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toMatchSnapshot();
	});

	it('renders a footer component with buttons', () => {
		const ModalWithState = () => {
			const {observer, onClose} = useModal({onClose: () => {}});
			return (
				<ClayModal observer={observer} spritemap={spritemap}>
					<ClayModal.Footer
						first={<Button>{'Bar'}</Button>}
						last={<Button onClick={onClose}>{'Foo'}</Button>}
						middle={<Button>{'Baz'}</Button>}
					/>
				</ClayModal>
			);
		};

		render(<ModalWithState />);

		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toMatchSnapshot();
	});
});
