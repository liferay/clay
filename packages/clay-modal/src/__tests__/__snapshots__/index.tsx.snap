// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayModal renders 1`] = `
<body
  class="modal-open"
>
  <div>
    <div
      class="modal-backdrop fade show"
    />
    <div
      class="fade modal d-block show"
    >
      <div
        class=""
        style="margin: auto; max-width: 600px;"
      >
        <div
          class="modal-dialog"
          tabindex="-1"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <div
                class="modal-title"
              >
                Foo
              </div>
              <button
                aria-label="close"
                class="close btn btn-unstyled"
                type="button"
              >
                <svg
                  class="lexicon-icon lexicon-icon-times"
                  role="presentation"
                >
                  <use
                    xlink:href="icons.svg#times"
                  />
                </svg>
              </button>
            </div>
            <div
              class="modal-body"
            >
              <h1>
                Hello world!
              </h1>
            </div>
            <div
              class="modal-footer"
            >
              <div
                class="modal-item-first"
              />
              <div
                class="modal-item"
              />
              <div
                class="modal-item-last"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Primary
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`ClayModal renders Header w/ low-level API 1`] = `
<body
  class="modal-open"
>
  <div>
    <div
      class="modal-backdrop fade show"
    />
    <div
      class="fade modal d-block show"
    >
      <div
        class=""
        style="margin: auto; max-width: 600px;"
      >
        <div
          class="modal-dialog"
          tabindex="-1"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <div
                class="modal-item-group"
              >
                <div
                  class="modal-item"
                >
                  <div
                    class="modal-title-section"
                  >
                    <div
                      class="modal-title"
                    >
                      Modal Title
                    </div>
                  </div>
                </div>
                <div
                  class="modal-item modal-item-shrink"
                >
                  <div
                    class="modal-subtitle-section"
                  >
                    <div
                      class="modal-subtitle"
                    >
                      Modal Subtitle
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`ClayModal renders a body component with url 1`] = `
<body
  class="modal-open"
>
  <div>
    <div
      class="modal-backdrop fade show"
    />
    <div
      class="fade modal d-block show"
    >
      <div
        class=""
        style="margin: auto; max-width: 600px;"
      >
        <div
          class="modal-dialog"
          tabindex="-1"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-body modal-body-iframe"
            >
              <iframe
                src="http://localhost"
                title="http://localhost"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`ClayModal renders a footer component with buttons 1`] = `
<body
  class="modal-open"
>
  <div>
    <div
      class="modal-backdrop fade show"
    />
    <div
      class="fade modal d-block show"
    >
      <div
        class=""
        style="margin: auto; max-width: 600px;"
      >
        <div
          class="modal-dialog"
          tabindex="-1"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-footer"
            >
              <div
                class="modal-item-first"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Bar
                </button>
              </div>
              <div
                class="modal-item"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Baz
                </button>
              </div>
              <div
                class="modal-item-last"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Foo
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`ClayModal renders with Header 1`] = `
<body
  class="modal-open"
>
  <div>
    <div
      class="modal-backdrop fade show"
    />
    <div
      class="fade modal d-block show"
    >
      <div
        class=""
        style="margin: auto; max-width: 600px;"
      >
        <div
          class="modal-dialog"
          tabindex="-1"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <div
                class="modal-title"
              >
                Foo
              </div>
              <button
                aria-label="close"
                class="close btn btn-unstyled"
                type="button"
              >
                <svg
                  class="lexicon-icon lexicon-icon-times"
                  role="presentation"
                >
                  <use
                    xlink:href="icons.svg#times"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`ClayModal renders with size 1`] = `
<body
  class="modal-open"
>
  <div>
    <div
      class="modal-backdrop fade show"
    />
    <div
      class="fade modal d-block show"
    >
      <div
        class="modal-lg"
        style="margin: auto;"
      >
        <div
          class="modal-dialog modal-lg"
          tabindex="-1"
        >
          <div
            class="modal-content"
          />
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`ClayModal renders with status 1`] = `
<body
  class="modal-open"
>
  <div>
    <div
      class="modal-backdrop fade show"
    />
    <div
      class="fade modal d-block show"
    >
      <div
        class=""
        style="margin: auto; max-width: 600px;"
      >
        <div
          class="modal-dialog modal-success"
          tabindex="-1"
        >
          <div
            class="modal-content"
          />
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;
